File: project/CCNet.VBDemos.Plugin/SourceControl/NewFileSystemSourceControl.vb
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports Exortech.NetReflector
Imports ThoughtWorks.CruiseControl.Core

<ReflectorType("fileSystemSource")> Public Class NewFileSystemSourceControl
    Implements ISourceControl

    Private myFiles As Dictionary(Of String, Date) = New Dictionary(Of String, Date)()

    Private myDirectoryName As String
    <ReflectorProperty("directory")> _
    Public Property DirectoryName() As String
        Get
            Return myDirectoryName
        End Get
        Set(ByVal value As String)
            myDirectoryName = value
        End Set
    End Property

    Public Function GetModifications(ByVal from As IIntegrationResult, _
                                     ByVal [to] As IIntegrationResult) As Modification() _
                                     Implements ISourceControl.GetModifications
        Dim newList = New Dictionary(Of String, DateTime)()
        Dim modifications = New List(Of Modification)()
        Dim directory = New DirectoryInfo(DirectoryName)
        Dim newFiles = directory.GetFiles()
        For Each file In newFiles
            Dim inList = myFiles.ContainsKey(file.FullName)
            If (Not inList Or (myFiles(file.FullName) < file.LastWriteTime)) Then
                newList.Add(file.FullName, file.LastWriteTime)
                Dim newModification = New Modification()
                newModification.FileName = file.Name
                newModification.ModifiedTime = file.LastWriteTime
                newModification.FolderName = file.DirectoryName
                If inList Then
                    newModification.Type = "Added"
                Else
                    newModification.Type = "Modified"
                End If

                modifications.Add(newModification)
            End If

                If (inList) Then
                myFiles.Remove(file.Name)
            End If
        Next

        For Each file In myFiles.Keys
            Dim oldModification = New Modification()
            oldModification.FileName = Path.GetFileName(file)
            oldModification.ModifiedTime = Date.Now
            oldModification.FolderName = Path.GetDirectoryName(file)
            oldModification.Type = "Deleted"
            modifications.Add(oldModification)
        Next

        myFiles = newList
        Return modifications.ToArray()
    End Function

    Public Sub GetSource(ByVal result As IIntegrationResult) _
    Implements ISourceControl.GetSource
        For Each Modification In result.Modifications
            Dim source = Path.Combine(Modification.FolderName, _
                Modification.FileName)
            Dim destination = result.BaseFromWorkingDirectory(Modification.FileName)
            If (File.Exists(source)) Then
                File.Copy(source, destination, True)
            Else
                File.Delete(destination)
            End If
        Next
    End Sub

    Public Sub Initialize(ByVal project As IProject) _
    Implements ISourceControl.Initialize

    End Sub

    Public Sub LabelSourceControl(ByVal result As IIntegrationResult) _
    Implements ISourceControl.LabelSourceControl
        Dim fileName = Path.Combine(myDirectoryName, _
                                    Date.Now.ToString("yyyyMMddHHmmss") & ".label")
        File.WriteAllText(fileName, result.Label)
        myFiles.Add(fileName, DateTime.Now)
    End Sub

    Public Sub Purge(ByVal project As IProject) _
    Implements ISourceControl.Purge

    End Sub
End Class

File: project/CCNet.CSharpDemos.Plugin/SourceControl/NewFileSystemSourceControl.cs
namespace CCNet.CSharpDemos.Plugin.SourceControl
{
    using System;
    using System.Collections.Generic;
    using System.IO;
    using Exortech.NetReflector;
    using ThoughtWorks.CruiseControl.Core;

    [ReflectorType("fileSystemSource")]
    public class NewFileSystemSourceControl
        : ISourceControl
    {
        private Dictionary<string, DateTime> files = new Dictionary<string, DateTime>();

        [ReflectorProperty("directory")]
        public string DirectoryName { get; set; }

        public Modification[] GetModifications(IIntegrationResult from, IIntegrationResult to)
        {
            var newList = new Dictionary<string, DateTime>();
            var modifications = new List<Modification>();
            var directory = new DirectoryInfo(this.DirectoryName);
            var newFiles = directory.GetFiles();
            foreach (var file in newFiles)
            {
                var inList = this.files.ContainsKey(file.FullName);
                if (!inList ||
                    (this.files[file.FullName] < file.LastWriteTime))
                {
                    newList.Add(file.FullName, file.LastWriteTime);
                    var modification = new Modification
                    {
                        FileName = file.Name,
                        ModifiedTime = file.LastWriteTime,
                        FolderName = file.DirectoryName,
                        Type = inList ? "Added" : "Modified"
                    };
                    modifications.Add(modification);
                }

                if (inList)
                {
                    this.files.Remove(file.Name);
                }
            }

            foreach (var file in this.files.Keys)
            {
                var modification = new Modification
                {
                    FileName = Path.GetFileName(file),
                    ModifiedTime = DateTime.Now,
                    FolderName = Path.GetDirectoryName(file),
                    Type = "Deleted"
                };
                modifications.Add(modification);
            }

            this.files = newList;
            return modifications.ToArray();
        }

        public void LabelSourceControl(IIntegrationResult result)
        {
            var fileName = Path.Combine(
                this.DirectoryName,
                DateTime.Now.ToString("yyyyMMddHHmmss") + ".label");
            File.WriteAllText(fileName, result.Label);
            this.files.Add(fileName, DateTime.Now);
        }

        public void GetSource(IIntegrationResult result)
        {
            foreach (var modification in result.Modifications)
            {
                var source = Path.Combine(
                    modification.FolderName,
                    modification.FileName);
                var destination = result.BaseFromWorkingDirectory(
                    modification.FileName);
                if (File.Exists(source))
                {
                    File.Copy(source, destination, true);
                }
                else
                {
                    File.Delete(destination);
                }
            }
        }

        public void Initialize(IProject project)
        {
            // Not needed
        }

        public void Purge(IProject project)
        {
            // Not needed
        }
    }
}
